### PR TITLE
Initial PR for VSINPU execution provider

### PR DESCRIPTION
**Pull request by chenfeiyue-cfy**
### Description

-It is an initial PR for VSINPU execution provider

### Motivation and Context

- For support VeriSilicon hardware
- TIM-VX(Tensor Interface Module) (https://github.com/VeriSilicon/TIM-VX) is an integrated software solution by Verisilicon for our hardware(A311D/i.MX 8M Plus etc.) design, it is easy to use Verisilicon’s hardware by simply connecting onnxruntime with the TIM-VX API by this VSINPU execution provider.

### CODE REVIEW

**Comment by chenfeiyue-cfy** [comment]
@sunshinemyson @thezha Could you please confirm if we can sign this CLA？

**Comment by chenfeiyue-cfy** [comment]
@microsoft-github-policy-service agree company="VeriSilicon"

File: onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_cosine_sim.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import sys
+import numpy as np
+from numpy.linalg import norm
+
+def read_values(filename):
+    with open(filename, 'r') as file:
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_cosine_sim.py]: done

File: onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_cosine_sim.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import sys
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_cosine_sim.py:1]: done

[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_cosine_sim.py:1]: done

File: onnxruntime/core/providers/vsinpu/builders/impl/reduce_op_builder.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/builders/impl/reduce_op_builder.h:1]: done

File: include/onnxruntime/core/providers/vsinpu/vsinpu_provider_factory.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply include/onnxruntime/core/providers/vsinpu/vsinpu_provider_factory.h:1]: done

File: onnxruntime/core/providers/vsinpu/builders/impl/clip_op_builder.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/builders/impl/clip_op_builder.h:1]: done

File: onnxruntime/core/providers/vsinpu/builders/impl/norm_op_builder.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/builders/impl/norm_op_builder.h:1]: done

File: onnxruntime/core/providers/vsinpu/builders/impl/base_op_builder.cc [hunk 7]
```diff
@@ -0,0 +1,204 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/builders/impl/base_op_builder.cc:1]: done

File: onnxruntime/core/providers/vsinpu/builders/impl/dequantize_op_builder.h [hunk 8]
```diff
@@ -0,0 +1,80 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/builders/impl/dequantize_op_builder.h:1]: done

File: onnxruntime/core/providers/vsinpu/builders/impl/base_op_builder.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/builders/impl/base_op_builder.cc:1]: done

File: onnxruntime/core/providers/vsinpu/builders/op_builder_factory.h [hunk 9]
```diff
@@ -0,0 +1,130 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/builders/op_builder_factory.h:1]: done

File: onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_topn.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import sys
+
+def read_values(filename):
+    with open(filename, 'r') as file:
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_topn.py]: done

File: onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_topn.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import sys
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_topn.py:1]: done

[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/patches/test_scripts/compare_topn.py:1]: done

File: onnxruntime/core/providers/vsinpu/vsinpu_ep_graph.h [hunk 12]
```diff
@@ -0,0 +1,114 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/vsinpu_ep_graph.h:1]: done

File: onnxruntime/core/providers/vsinpu/vsinpu_util.cc [hunk 13]
```diff
@@ -0,0 +1,501 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (chenfeiyue-cfy) [reply onnxruntime/core/providers/vsinpu/vsinpu_util.cc:1]: done

File: onnxruntime/core/framework/node_unit.cc [hunk 14]
```diff
@@ -351,6 +351,13 @@ void NodeUnit::InitForSingleNode() {
                                      NodeUnitIODef::QuantParam{*input_defs[1],
                                                                input_defs.size() == 3 ? input_defs[2] : nullptr,
                                                                axis}});
+  } else if (IsVariadicQLinearOp(qlinear_type)) {
```
Comment (jywu-msft) [inline onnxruntime/core/framework/node_unit.cc:354]: @skottmckay, can you take a quick look at this newly added code in this file?

Comment (skottmckay) [reply onnxruntime/core/framework/node_unit.cc:354]: LGTM

Comment (jywu-msft) [reply onnxruntime/core/framework/node_unit.cc:354]: thanks!

**Comment by jywu-msft** [comment]
can you provide some details about the patch files under vsinpu/patches? why are they there?

**Comment by chenfeiyue-cfy** [comment]
> can you provide some details about the patch files under vsinpu/patches? why are they there?

Sure! There are three patches under this directory.I will introduce the function of each patch file separately.

- AccuracyCorrection.patch()：This patch is used to address the precision discrepancy (with a difference of 1) in the uint8 data type.
- local_testing_record_res.patch(): It is a patch of workaround method. The purpose of this patch is to temporarily and informally record each output in order, making it easier to compare the correctness and similarity between our hardware results and the golden values using other scripts later.
- mlas_crosscompiling.patch(): We need to temporary disable mlas_fp16 under our ARM64 HWs, because there exists conflict of compiliation flag when cross compiling the project. This patch is only needed on our some Specific HW.

**Review by jywu-msft — APPROVED** [review]

**Comment by jiangzhiwei2018** [comment]
hi @chenfeiyue-cfy , I‘m tring to build onnxruntiom from soure code with `--use_vsinpu` .
What's the minimum version of [TIM-VX](https://github.com/VeriSilicon/TIM-VX) is required ?

**Pull request merged**